torch_geometric/nn/models/lightgcn.py

```diff
@@ -0,0 +1,315 @@
+from torch_geometric.typing import Adj
+
+import torch
+from torch import Tensor, LongTensor
+import torch.nn.functional as F
+from torch.nn import ModuleList, Embedding, BCELoss
+from torch.nn.modules.loss import _Loss
+
+from torch_geometric.nn.conv import LGConv
+
+
+class LightGCN(torch.nn.Module):
+    r"""The LightGCN model from the `"LightGCN: Simplifying and Powering
+    Graph Convolution Network for Recommendation"
+    <https://arxiv.org/abs/2002.02126>`_ paper.
+
+    LightGCN learns embeddings by linearly propagating them on the underlying
+    graph, and uses the weighted sum of the embeddings learned at all layers as
+    the final embedding as follows:
+
+    .. math::
+        \textbf{x}_i = \sum_{l=0}^{L} \alpha_l \textbf{x}^{l}_i
+
+    Each layer's embeddings are computed as:
+
+    .. math::
+        \mathbf{x}^{l+1}_i = \sum_{j \in \mathcal{N}(i)}
+        \frac{1}{\sqrt{\deg(i)}\sqrt{\deg(j)}}\mathbf{x}^{l}_j
+
+    Two prediction heads are provided: :obj:`"ranking"` and
+    :obj:`"link_prediction"`. Ranking is performed by outputting the dot
+    product between the pair of embeddings chosen for evaluation. For link
+    prediction, the sigmoid of the ranking is computed to output label
+    probabilities.
+
+    .. note::
+
+        This implementation works not only on bipartite graphs but on any
+        kind of connected graph. Embeddings are propagated according to the
+        graph connectivity specified by :obj:`edge_index` while rankings or
+        probabilities are only computed (the forward pass) according to the
+        edges specified by :obj:`edge_label_index`.
+
+    Args:
+        num_layers (int): The number of Light Graph Convolution
+            (:obj:`LGConv`) layers.
+        num_nodes (int): The number of nodes in the graph.
+        embedding_dim (int, optional): The dimensionality of the node
+            embeddings (default: 32).
+        objective (str, optional): The objective of the model. The original
```

Comment:
Shall we align this with `GAE` (https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/autoencoder.py#L26-L28), e.g.:
```suggestion
        sigmoid (bool, optional): ...
```

Reply:
This would be nice, but I think having the "objective" helps to enforce the right model usage + help the developer decide which kind of model to use depending on the task (i.e. choosing "recommendation_loss" vs. "link_prediction_loss"). 

Eventually, it would be cool if people could implement their own "prediction head" classes to stack on top of a model depending on the objective (node or edge classification / regression, link prediction, ranking etc.) similar to how models make use of convolutional layers in /nn/conv it would be nice to have a /nn/heads directory which implemented common model prediction heads. This would essentially be an abstraction of the what is on slide 24 here: [CS224W Stanford](http://snap.stanford.edu/class/cs224w-2020/slides/08-GNN-application.pdf). Would this work for you?

Reply:
Yes, I absolutely love this idea, in which we provide prediction heads for link-level or graph-level tasks.